runtime/lua/vim/lsp/semantic_tokens.lua

```diff
@@ -637,23 +651,58 @@ end
 --- Only has an effect if the buffer is currently active for semantic token
 --- highlighting (|vim.lsp.semantic_tokens.start()| has been called for it)
 ---
----@param bufnr (nil|number) default: current buffer
+---@param bufnr (number|nil) filter by buffer. All buffers if nil, current
+---       buffer if 0
 function M.force_refresh(bufnr)
   vim.validate({
     bufnr = { bufnr, 'n', true },
   })
 
-  if bufnr == nil or bufnr == 0 then
-    bufnr = api.nvim_get_current_buf()
+  local buffers = bufnr == nil and vim.tbl_keys(STHighlighter.active)
+    or bufnr == 0 and { api.nvim_get_current_buf() }
+    or { bufnr }
+
+  for _, buffer in ipairs(buffers) do
+    local highlighter = STHighlighter.active[buffer]
+    if highlighter then
+      highlighter:reset()
+      highlighter:send_request()
+    end
   end
+end
 
+--- Highlight a semantic token by applying an extmark. This extmark will be
+--- deleted by the semantic token decoration provider on the next `on_win`
+--- update.
+---
+---@param token (table) a semantic token as returned by
+---       |vim.lsp.semantic_tokens.get_at_pos()|
+---@param bufnr (number) the buffer to highlight
+---@param client_id (number) The ID of the |vim.lsp.client|
+---@param hl_group (string) Highlight group name
+---@param opts (table|nil) Optional parameters.
+---       - priority: (number|nil) Priority for the applied extmark. Defaults
+---         to `vim.highlight.priorities.semantic_tokens + 3`
+function M.highlight_token(token, bufnr, client_id, hl_group, opts)
   local highlighter = STHighlighter.active[bufnr]
   if not highlighter then
     return
   end
 
-  highlighter:reset()
-  highlighter:send_request()
+  local state = highlighter.client_state[client_id]
+  if not state then
```

Comment:
Probably also need to check that the buffer is active for the client.

Reply:
Can you explain further? My (likely flawed) thought was:

- Get the highlighter for the buffer.
- Get the state for the given client from the highlighter.

I was thinking that, if the buffer is not active for a client, one of those would be `nil`.

Reply:
I have no idea what I was talking about. I only saw one of the checks and my brain said one was missing. Ignore me.

Reply:
No problem!